File: ad-hoc/UVa100-the-3n-plus-1-problem.py

def cycle_length(n):

    def cycle(n, param_count):

        count = param_count + 1
        # print n,

        if n != 1:
            if n % 2 != 0:
                n = 3 * n + 1
            else:
                n = n / 2

            return cycle(n, count)
        else:
            return count

    return cycle(n, 0)


def max_cycle_length(first, second):
    
    max_value = -1

    if first > second:
        first, second = second, first

    iterator_number = first

    while(first <= second):
        current_cycle_length = cycle_length(first)

        if current_cycle_length > max_value:
            max_value = current_cycle_length

        first = first + 1

    return "%d, %d, %d" %  (iterator_number, second, max_value)

print max_cycle_length(1, 10)
# print "count is ", count

File: ad-hoc/uva100-the-3n-plus-1-problem.cpp
#include <iostream>
#include <stdio.h>

using namespace std;

int cycle_length(int x){
	int length = 1;
	while( x!=1 ){
		if ( x%2 == 0 ){
			x = x/2;
		}else{
			x = 3*x + 1;
		}
		length++;
	}
	return length;
}

int max_cycle_length(int a, int b){
	if(a>b){
		int temp = a;
		a = b;
		b = temp;
	}

	int max_length = 0;

	while(a<=b){
		int current_cycle_length = cycle_length(a);
		if (current_cycle_length > max_length){
			max_length = current_cycle_length;
		}
	}

	return max_length;
}

int main(){
	int a, b;

	while (scanf("%d %d", &a, &b) != EOF) {
		printf("%d %d %d\n", a, b, maxCycleSizeBetween(a, b));
    }
	
	return 0;
}

